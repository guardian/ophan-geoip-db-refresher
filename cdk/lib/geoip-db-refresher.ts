import { GuScheduledLambda } from '@guardian/cdk';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { Schedule } from 'aws-cdk-lib/aws-events';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Architecture, Runtime } from 'aws-cdk-lib/aws-lambda';

export class GeoipDbRefresher extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const app = 'geoip-db-refresher';

		const loggingPolicy = new PolicyStatement({
			resources: ['arn:aws:logs:*:*:*'],
			actions: [
				'logs:CreateLogGroup',
				'logs:CreateLogStream',
				'logs:PutLogEvents',
			],
		});

		const geoIpDbRefresherActionsPolicy = new PolicyStatement({
			resources: ['arn:aws:s3:::ophan-dist/geoip/*'],
			actions: ['s3:PutObject', 's3:PutObjectAcl'],
		});

		const ssmGetParameterPolicy = new PolicyStatement({
			resources: [
				`arn:aws:ssm:eu-west-1:${this.account}:parameter/Ophan/GeoIP`,
			],
			actions: ['ssm:GetParameter'],
		});

		const kmsDecryptPolicy = new PolicyStatement({
			resources: [
				`arn:aws:kms:eu-west-1:${this.account}:key/d77985cc-fb91-42e5-86f9-505fe2eefb76`,
			],
			actions: ['kms:Decrypt'],
		});

		const lambda = new GuScheduledLambda(this, 'geoip-db-refresher', {
			app,
			fileName: 'geoip-db-refresher.jar',
			description:
				'Fetching the latest GeoIP database and putting it in S3 for Ophan',
			handler: 'ophan.geoip.db.refresher.Lambda::handler',
			runtime: Runtime.JAVA_11,
			memorySize: 1536, // more memory than we need, but we're billed for fewer GB-seconds this way
			architecture: Architecture.ARM_64,
			timeout: Duration.seconds(120),
			// MaxMind: "Databases updated twice-weekly on Tuesdays and Fridays" ... can be "delayed by about one day"
			rules: [
				{ schedule: Schedule.expression('cron(20 11 ? * WED,THU,SAT,SUN *)') },
			],
			monitoringConfiguration: { noMonitoring: true },
		});

		lambda.addToRolePolicy(loggingPolicy);
		lambda.addToRolePolicy(geoIpDbRefresherActionsPolicy);
		lambda.addToRolePolicy(ssmGetParameterPolicy);
		lambda.addToRolePolicy(kmsDecryptPolicy);
	}
}
