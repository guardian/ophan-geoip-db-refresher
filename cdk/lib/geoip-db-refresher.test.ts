import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { GeoipDbRefresher } from "./geoip-db-refresher";

describe("The GeoipDbRefresher stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new GeoipDbRefresher(app, "GeoipDbRefresher", { stack: "ophan", stage: "TEST" });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
