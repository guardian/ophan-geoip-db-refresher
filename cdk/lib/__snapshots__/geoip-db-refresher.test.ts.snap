// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GeoipDbRefresher stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "geoipdbrefresherABBF7BA3": {
      "DependsOn": [
        "geoipdbrefresherServiceRoleDefaultPolicy86D5F709",
        "geoipdbrefresherServiceRole2ABAFA67",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "ophan/TEST/geoip-db-refresher/geoip-db-refresher.jar",
        },
        "Description": "Fetching the latest GeoIP database and putting it in S3 for Ophan",
        "Environment": {
          "Variables": {
            "APP": "geoip-db-refresher",
            "STACK": "ophan",
            "STAGE": "TEST",
          },
        },
        "Handler": "ophan.geoip.db.refresher.Lambda::handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "geoipdbrefresherServiceRole2ABAFA67",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": [
          {
            "Key": "App",
            "Value": "geoip-db-refresher",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/ophan-geoip-db-refresher",
          },
          {
            "Key": "Stack",
            "Value": "ophan",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "geoipdbrefresherServiceRole2ABAFA67": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "geoip-db-refresher",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/ophan-geoip-db-refresher",
          },
          {
            "Key": "Stack",
            "Value": "ophan",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "geoipdbrefresherServiceRoleDefaultPolicy86D5F709": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/ophan/TEST/geoip-db-refresher/geoip-db-refresher.jar",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/ophan/geoip-db-refresher",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/ophan/geoip-db-refresher/*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:logs:*:*:*",
            },
            {
              "Action": [
                "s3:PutObject",
                "s3:PutObjectAcl",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::ophan-dist/geoip/*",
            },
            {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/Ophan/GeoIP",
                  ],
                ],
              },
            },
            {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:kms:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":key/d77985cc-fb91-42e5-86f9-505fe2eefb76",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "geoipdbrefresherServiceRoleDefaultPolicy86D5F709",
        "Roles": [
          {
            "Ref": "geoipdbrefresherServiceRole2ABAFA67",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "geoipdbrefreshergeoipdbrefreshercron2011WEDTHUSATSUN02C2AAFFB": {
      "Properties": {
        "ScheduleExpression": "cron(20 11 ? * WED,THU,SAT,SUN *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "geoipdbrefresherABBF7BA3",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "geoipdbrefreshergeoipdbrefreshercron2011WEDTHUSATSUN0AllowEventRuleGeoipDbRefreshergeoipdbrefresher75E55D0C16B93CAF": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "geoipdbrefresherABBF7BA3",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "geoipdbrefreshergeoipdbrefreshercron2011WEDTHUSATSUN02C2AAFFB",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
